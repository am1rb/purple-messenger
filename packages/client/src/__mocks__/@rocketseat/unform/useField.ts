const useField = jest.fn();

export default useField;
