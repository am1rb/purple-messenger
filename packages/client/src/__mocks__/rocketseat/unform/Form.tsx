import React from "react";

interface FormProps {
  children: React.ReactNode;
}

function Form({ children }: FormProps) {
  return <form data-testid="mock-form">{children}</form>;
}

export default Form;
