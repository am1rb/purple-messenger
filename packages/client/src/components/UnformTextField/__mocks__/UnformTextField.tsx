import React from "react";

interface TextAreaProps extends React.HTMLAttributes<HTMLTextAreaElement> {
  multiline: true;
}

type InputProps = React.HTMLAttributes<HTMLInputElement>;

type TagProps = TextAreaProps & InputProps;

interface UnformTextFieldProps extends TagProps {
  fullWidth?: boolean;
}

const unformTextField = jest.fn(
  ({ multiline, fullWidth, ...other }: UnformTextFieldProps) =>
    multiline ? (
      <textarea data-testid="mock-unform-text-field" {...other} />
    ) : (
      <input data-testid="mock-unform-text-field" {...other} />
    )
);

export default unformTextField;
