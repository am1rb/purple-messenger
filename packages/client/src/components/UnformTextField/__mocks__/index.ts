export { default } from "./UnformTextField";
export * from "./UnformTextField";
