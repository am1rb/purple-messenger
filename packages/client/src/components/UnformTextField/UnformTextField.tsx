import React, { useRef, useEffect } from "react";
import TextField, {
  TextFieldProps,
  StandardTextFieldProps,
} from "@material-ui/core/TextField";
import { useField } from "@rocketseat/unform";

export type UnformTextFieldProps = Omit<
  TextFieldProps,
  "name" | "inputRef" | "defaultValue" | "defaultChecked"
> & {
  name: string;
};

function UnformTextField({ name, helperText, ...other }: UnformTextFieldProps) {
  const ref = useRef<HTMLInputElement>(null);
  const { fieldName, registerField, defaultValue, error } = useField(name);

  useEffect(() => {
    registerField({
      name: fieldName,
      ref: ref.current as HTMLInputElement,
      path: "value",
      clearValue: () => {
        ref.current!.value = "";
      },
    });
  }, [fieldName]); // eslint-disable-line

  return (
    <TextField
      inputRef={ref}
      defaultValue={defaultValue}
      error={!!error}
      helperText={error || helperText}
      {...(other as StandardTextFieldProps)}
    />
  );
}

export default UnformTextField;
