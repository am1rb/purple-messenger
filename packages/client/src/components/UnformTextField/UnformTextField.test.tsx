import { render } from "@testing-library/react";
import React from "react";
import { useField } from "@rocketseat/unform";
import UnformTextField from "./UnformTextField";

jest.mock("@rocketseat/unform");

describe("UnformTextField Component Tests", () => {
  it("Should match the snapshot when there is a simple input", () => {
    const { container } = render(<UnformTextField name="myField" />);
    expect(container).toMatchSnapshot();
  });

  it("Should match the snapshot when the field has a label", () => {
    const { container } = render(
      <UnformTextField name="myField" label="field label" />
    );
    expect(container).toMatchSnapshot();
  });

  it("Should match the snapshot when the field has a helper text", () => {
    const { container } = render(
      <UnformTextField name="myField" helperText="helper text" />
    );
    expect(container).toMatchSnapshot();
  });

  it("should register the field in unform correctly", () => {
    const registerField = jest.fn();
    const useFieldMock = (useField as jest.Mock).mockReturnValue({
      registerField,
      fieldName: "customFieldName",
    });
    render(<UnformTextField name="myField" />);
    expect(registerField).toHaveBeenCalledWith({
      name: "customFieldName",
      clearValue: expect.anything(),
      path: "value",
      ref: expect.anything(),
    });
    useFieldMock.mockClear();
  });

  it("should render the field error correctly", () => {
    const errorMessage = "field error message";
    const useFieldMock = (useField as jest.Mock).mockReturnValue({
      registerField: () => {},
      fieldName: "field",
      error: errorMessage,
    });
    const { getByTestId } = render(<UnformTextField name="myField" />);
    expect(getByTestId("mock-mui-text-field-helper-text")).toHaveTextContent(
      errorMessage
    );
    useFieldMock.mockClear();
  });

  it("should use the default value properly", () => {
    const defaultValue = "hello world";
    const useFieldMock = (useField as jest.Mock).mockReturnValue({
      registerField: () => {},
      fieldName: "field",
      defaultValue,
    });
    const { getByRole } = render(<UnformTextField name="myField" />);
    expect(getByRole("textbox")).toHaveValue(defaultValue);
    useFieldMock.mockClear();
  });
});
