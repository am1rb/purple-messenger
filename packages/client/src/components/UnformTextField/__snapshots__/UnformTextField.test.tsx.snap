// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnformTextField Component Tests Should match the snapshot when the field has a helper text 1`] = `
<div>
  <input
    data-testid="mock-mui-text-field"
    type="text"
    value="defaultValue"
  />
  <span
    data-testid="mock-mui-text-field-helper-text"
  >
    helper text
  </span>
</div>
`;

exports[`UnformTextField Component Tests Should match the snapshot when the field has a label 1`] = `
<div>
  <label>
    <span
      data-testid="mock-mui-text-field-label"
    >
      field label
    </span>
    <input
      data-testid="mock-mui-text-field"
      type="text"
      value="defaultValue"
    />
  </label>
</div>
`;

exports[`UnformTextField Component Tests Should match the snapshot when there is a simple input 1`] = `
<div>
  <input
    data-testid="mock-mui-text-field"
    type="text"
    value="defaultValue"
  />
</div>
`;
